tools/syz-trace2syz/proggen/generate_unions.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package proggen
+
+import (
+	"strings"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/tools/syz-trace2syz/parser"
+)
+
+func genIpv4Addr(syzType *prog.UnionType, irType parser.IrType, ctx *Context) prog.Arg {
+	if syzType.Dir() == prog.DirOut {
+		return prog.DefaultArg(syzType)
+	}
+	var ip uint64
+	switch a := irType.(type) {
+	case parser.Expression:
+		ip = a.Eval(ctx.Target)
+	default:
+		log.Fatalf("Failed to parse ip4addr. Expected Expression irtype got: %#v", a)
+	}
+	for _, field := range syzType.Fields {
+		if !strings.Contains(field.FieldName(), "rand") {
```

Comment:
ipv4_addr does not have option "rand", it has:
```
	rand_addr	int32be
```
So this never worked.
1. We need tests that all functions in this file and other similar functions actually work.
2. We need to reorder union fields so that this field is the first one. And the same for most unions in this file. And then just use Fields[0] without any special casing and 200 lines of custom code.

Reply:
Wait no this actually did work since "rand" is a substring of "rand_addr"! Also, I do have a test [here](https://github.com/shankarapailoor/syzkaller/blob/master/tools/syz-trace2syz/proggen/proggen_test.go#L318)

Reply:
Then sorry.